// tower/test/example/public/javascripts/config/assets.js
var File, testCases, testModels;

File = require('pathfinder').File;

testCases = _.map(File.files("" + Tower.root + "/public/javascripts/test/cases"), function(path) {
  return path.replace("" + Tower.root + "/public/javascripts", "").replace(/\.js$/, "");
});

testCases = _.select(testCases, function(path) {
  return path.match(/model|application|store|support|http|client|controller/);
});

testModels = _.map(File.files("" + Tower.root + "/public/javascripts/app"), function(path) {
  return path.replace("" + Tower.root + "/public/javascripts", "").replace(/\.js$/, "");
});

testModels = _.select(testModels, function(path) {
  return path.match(/model|application|controller/) && !path.match('client');
});

module.exports = {
  javascripts: {
    application: ["/app/client/config/application", "/config/routes", "/app/views/templates"].concat(testModels),
    lib: [],
    vendor: ["/vendor/javascripts/underscore", "/vendor/javascripts/underscore.string", "/vendor/javascripts/moment", "/vendor/javascripts/geolib", "/vendor/javascripts/validator", "/vendor/javascripts/accounting", "/vendor/javascripts/inflection", "/vendor/javascripts/coffeekup", "/vendor/javascripts/prettify", "/vendor/javascripts/async", "/vendor/javascripts/socket.io", "/vendor/javascripts/history", "/vendor/javascripts/history.adapter.jquery", "/vendor/javascripts/bootstrap/bootstrap-transition", "/vendor/javascripts/bootstrap/bootstrap-alert", "/vendor/javascripts/bootstrap/bootstrap-modal", "/vendor/javascripts/bootstrap/bootstrap-dropdown", "/vendor/javascripts/bootstrap/bootstrap-scrollspy", "/vendor/javascripts/bootstrap/bootstrap-tab", "/vendor/javascripts/bootstrap/bootstrap-tooltip", "/vendor/javascripts/bootstrap/bootstrap-popover", "/vendor/javascripts/bootstrap/bootstrap-button", "/vendor/javascripts/bootstrap/bootstrap-collapse", "/vendor/javascripts/bootstrap/bootstrap-carousel", "/vendor/javascripts/bootstrap/bootstrap-typeahead", "/vendor/javascripts/ember", "/vendor/javascripts/tower"],
    development: ["/vendor/javascripts/mocha", "/vendor/javascripts/chai", "/vendor/javascripts/design.io", "/test/client"].concat(testCases)
  },
  stylesheets: {
    application: [],
    lib: [],
    vendor: [],
    development: ["/vendor/stylesheets/mocha"]
  }
};
