// docpad-plugin-coffee/out/coffee.plugin.js
// Generated by CoffeeScript 1.3.3
(function() {
  var __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  module.exports = function(BasePlugin) {
    var CoffeePlugin;
    return CoffeePlugin = (function(_super) {

      __extends(CoffeePlugin, _super);

      function CoffeePlugin() {
        return CoffeePlugin.__super__.constructor.apply(this, arguments);
      }

      CoffeePlugin.prototype.name = 'coffee';

      CoffeePlugin.prototype.priority = 700;

      CoffeePlugin.prototype.renderCoffeeKup = function(opts, next) {
        var ck, content, templateData;
        templateData = opts.templateData, content = opts.content;
        ck = require('coffeecup');
        opts.content = ck.render(content, templateData, this.config.coffeekup || this.config.coffeecup || {});
        return next();
      };

      CoffeePlugin.prototype.renderHTMLToCoffeeKup = function(opts, next) {
        var content, html2ck, outputStream, templateData;
        templateData = opts.templateData, content = opts.content;
        html2ck = require('html2coffeekup');
        outputStream = {
          content: '',
          write: function(data) {
            return this.content += data;
          }
        };
        return html2ck.convert(content, outputStream, function(err) {
          if (err) {
            return next(err);
          }
          opts.content = outputStream.content;
          return next();
        });
      };

      CoffeePlugin.prototype.renderCoffeeScriptToJavaScript = function(opts, next) {
        var coffee, content;
        content = opts.content;
        coffee = require('coffee-script');
        opts.content = coffee.compile(content);
        return next();
      };

      CoffeePlugin.prototype.renderJavaScriptToCoffeeScript = function(opts, next) {
        var content, js2coffee, templateData;
        templateData = opts.templateData, content = opts.content;
        js2coffee = require('js2coffee/lib/js2coffee.coffee');
        opts.content = js2coffee.build(content);
        return next();
      };

      CoffeePlugin.prototype.render = function(opts, next) {
        var inExtension, outExtension;
        inExtension = opts.inExtension, outExtension = opts.outExtension;
        if ((inExtension === 'coffeekup' || inExtension === 'coffeecup' || inExtension === 'ck') || (inExtension === 'coffee' && !(outExtension === 'js' || outExtension === 'css'))) {
          return this.renderCoffeeKup(opts, next);
        } else if (inExtension === 'html' && (outExtension === 'coffeekup' || outExtension === 'ck' || outExtension === 'coffee')) {
          return this.renderHTMLToCoffeeKup(opts, next);
        } else if ((inExtension === 'coffeescript' || inExtension === 'coffee') && (outExtension === 'js' || outExtension === null)) {
          return this.renderCoffeeScriptToJavaScript(opts, next);
        } else if (inExtension === 'js' && (outExtension === 'coffee' || outExtension === 'coffeescript')) {
          return this.renderJavaScriptToCoffeeScript(opts, next);
        } else {
          return next();
        }
      };

      return CoffeePlugin;

    })(BasePlugin);
  };

}).call(this);
